File: solutions/cpp/0394.cpp
class Solution {
 public:
  string decodeString(string s) {
    stack<string> stack;
    string currStr;
    int currNum = 0;

    for (const char c : s)
      if (isdigit(c)) {
        currNum = currNum * 10 + c - '0';
      } else if (c == '[') {
        stack.push(currStr);
        stack.push(to_string(currNum));
        currStr = "";
        currNum = 0;
      } else if (c == ']') {
        int repeatedTime = stoi(stack.top()); stack.pop();
        string repeatedStr;
        while (repeatedTime--)
          repeatedStr += currStr;
        string prevStr = stack.top(); stack.pop();
        currStr = prevStr + repeatedStr;
      } else {
        currStr += c;
      }

    return currStr;
  }
};

File: solutions/java/0394.java
class Solution {
  public String decodeString(String s) {
    String currStr = "";
    int currNum = 0;
    Stack<String> stack = new Stack<>();

    for (char c : s.toCharArray()) {
      if (Character.isDigit(c))
        currNum = currNum * 10 + c - '0';
      else if (c == '[') {
        stack.push(currStr);
        stack.push(String.valueOf(currNum));
        currStr = "";
        currNum = 0;
      } else if (c == ']') {
        int num = Integer.valueOf(stack.pop());
        String prevStr = stack.pop();
        String repeated = "";
        for (int i = 0; i < num; ++i)
          repeated += currStr;
        currStr = prevStr + repeated;
      } else
        currStr += c;
    }

    return currStr;
  }
}